broker/cloud_functions/tag/main.py

```diff
@@ -0,0 +1,266 @@
+ps_topic = f"{SURVEY}-tagged"
+
+if TESTID != "False":  # attach the testid to the names
+    bq_dataset = f"{bq_dataset}_{TESTID}"
+    ps_topic = f"{ps_topic}-{TESTID}"
+
+
+# Changed the name stub of the BigQuery table from SuperNNova to classifications
+class_table = f"{bq_dataset}.classifications"
+
+tags_table = f"{bq_dataset}.tags"
+
+
+
+
+def is_pure(alert_dict):
+    """Adapted from: https://zwickytransientfacility.github.io/ztf-avro-alert/filtering.html
+
+    Quoted from the source:
+
+    ZTF alert streams contain an nearly entirely unfiltered stream of all
+    5-sigma (only the most obvious artefacts are rejected). Depending on your
+    science case, you may wish to improve the purity of your sample by filtering
+    the data on the included attributes.
+
+    Based on tests done at IPAC (F. Masci, priv. comm), the following filter
+    delivers a relatively pure sample.
+    """
+
+    source = alert_dict['source']
+
+    rb = (source['rb'] >= 0.65)  # RealBogus score
+
+    if SURVEY == 'decat': ## How to get find survey type without schema_map
+        pure = rb
+
+    elif SURVEY == 'ztf': ## How to get find survey type without schema_map
+        nbad = (source['nbad'] == 0)  # num bad pixels
+        fwhm = (source['fwhm'] <= 5)  # Full Width Half Max, SExtractor [pixels]
+        elong = (source['elong'] <= 1.2)  # major / minor axis, SExtractor
+        magdiff = (abs(source['magdiff']) <= 0.1)  # aperture - psf [mag]
+        pure = (rb and nbad and fwhm and elong and magdiff)
+
+
+
+    purity_reason_dict = {
+        'is_pure': int(pure),
+        'rb': int(rb),
+        'nbad': int(nbad),
+        'fwhm': int(fwhm),
+        'elong': int(elong),
+        'magdiff': int(magdiff),
+    }
+
+    return purity_reason_dict
+
+
+
+def _is_extragalactic_transient(alert_dict: dict) -> dict:
+    """Check whether alert is likely to be an extragalactic transient.
+    Adapted from:
+    https://github.com/ZwickyTransientFacility/ztf-avro-alert/blob/master/notebooks/Filtering_alerts.ipynb
+    """
+    if SURVEY == "decat": ## How to get find survey type without schema_map
+        # No straightforward way to translate this ZTF filter for DECAT.
+        # DECAT alert does not include whether the subtraction (sci-ref) is
+        # positive, nor SExtractor results,
+        # and the included xmatch data is significantly different.
+        # However, DECAT is a transient survey.
+        # Assume the alert should pass the filter:
+        is_extragalactic_transient = True
+
+    elif SURVEY == "ztf": ## How to get find survey type without schema_map
+        dflc = alert_lite_to_dataframe(alert_dict) # Is schema_map necessary for this function?
+        # NEED TO GO TO data_utils.py and determine whether to take out schema map
+        
+        candidate = dflc.loc[0]
+
+        is_positive_sub = candidate["isdiffpos"] == "t"
+
+        if (candidate["distpsnr1"] is None) or (candidate["distpsnr1"] > 1.5):  # arcsec
```

Comment:
This seems to be the only place that `distpsnr1` is currently being used, and it only needs the value from one row of the dataframe (the one representing the most recent source/candidate observation). So let's just pull `distpsnr1` out of the `xmatch` dict right here instead of adding it to the dataframe. Try this:

```suggestion
        distpsnr1 = alert_dict["xmatch"]["distpsnr1"]
        if (distpsnr1 is None) or (distpsnr1 > 1.5):  # arcsec
```